src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryAddressImpl.java

```diff
@@ -68,16 +73,20 @@ public MemoryAddress addOffset(long offset) {
     public long segmentOffset(MemorySegment segment) {
         Objects.requireNonNull(segment);
         AbstractMemorySegmentImpl segmentImpl = (AbstractMemorySegmentImpl)segment;
-        if (segmentImpl.base() != base()) {
+        if (segmentImpl.base() != base() ||
+                (scope() != ResourceScope.globalScope() && segmentImpl.scope() != scope())) {
             throw new IllegalArgumentException("Invalid segment: " + segment);
         }
         return offset() - segmentImpl.min();
     }
 
     @Override
     public long toRawLongValue() {
-        if (base() != null) {
-            throw new UnsupportedOperationException("Not a native address");
+        if (segment != null) {
+            if (segment.base() != null) {
+                throw new UnsupportedOperationException("Not a native address");
```

Comment:
Maybe I'm wrong, but if I would like to deregister resource associated with this address here (I.e by calling unmap) during cleanup, will it cause exception?

Reply:
In general yes. Although I argue that if you get such an exception, the cleanup action is doing something suspicious. If you are referring to your allocator work, I think the trick there is to create entries that are backed by plain addresses (and where the addresses are the ones you obtain from malloc - so they have the global scope, no issue there).

But - if you find yourself in a tight spot during cleanup, you could always do:

```
long rawAddress = address.segmentOffset(MemorySegment.globalNativeScope);
MemoryAddress addr = MemoryAddress.ofLong(rawAddress);
```

Reply:
Yes, I was bit scared about it and one other thing I got in mind, but I think it will work. Thank you.